script.speedtester/default.py

```diff
@@ -0,0 +1,939 @@
+    return sock
+
+def distance(origin, destination):
+    (lat1, lon1) = origin
+    (lat2, lon2) = destination
+    radius = 6371  # km
+
+    dlat = math.radians(lat2 - lat1)
+    dlon = math.radians(lon2 - lon1)
+    a = math.sin(dlat / 2) * math.sin(dlat / 2) \
+        + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) \
+        * math.sin(dlon / 2) * math.sin(dlon / 2)
+    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
+    d = radius * c
+
+    return d
+
+def build_request(url, data=None, headers={}):
+    if url[0] == ':':
+        schemed_url = '%s%s' % (scheme, url)
+    else:
+        schemed_url = url
+    headers['User-Agent'] = user_agent
+    return Request(schemed_url, data=data, headers=headers)
+
+def catch_request(request):
+    try:
+        uh = urlopen(request)
+        return uh
+    except (HTTPError, URLError, socket.error):
+        e = sys.exc_info()[1]
+        return (None, e)
+
+class FileGetter(threading.Thread):
+
+    def __init__(self, url, start):
+        self.url = url
+        self.result = None
+        self.starttime = start
+        threading.Thread.__init__(self)
+
+    def run(self):
+        self.result = [0]
+        try:
+            if timeit.default_timer() - self.starttime <= 10:
+                request = build_request(self.url)
+                f = urlopen(request)
+                while 1 and not shutdown_event.isSet():
+                    self.result.append(len(f.read(10240)))
+                    if self.result[-1] == 0:
+                        break
+                f.close()
+        except IOError:
+            pass
+
+class FilePutter(threading.Thread):
+
+    def __init__(
+        self,
+        url,
+        start,
+        size,
+        ):
+        self.url = url
+        chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+        data = chars * int(round(int(size) / 36.0))
+        self.data = ('content1=%s' % data[0:int(size) - 9]).encode()
+        del data
+        self.result = None
+        self.starttime = start
+        threading.Thread.__init__(self)
+
+    def run(self):
+        try:
+            if timeit.default_timer() - self.starttime <= 10 \
+                and not shutdown_event.isSet():
+                request = build_request(self.url, data=self.data)
+                f = urlopen(request)
+                f.read(11)
+                f.close()
+                self.result = len(self.data)
+            else:
+                self.result = 0
+        except IOError:
+            self.result = 0
+
+def getAttributesByTagName(dom, tagName):
+    elem = dom.getElementsByTagName(tagName)[0]
+    return dict(list(elem.attributes.items()))
+
+def getConfig():
+    request = \
+        build_request('http://www.speedtest.net/speedtest-config.php')
+    uh = catch_request(request)
+    if uh is False:
+        xbmc.log('Could not retrieve speedtest.net configuration: %s' % e,xbmc.LOGDEBUG)
+        sys.exit(1)
+    configxml = []
+    while 1:
```

Comment:
Can you write the loop properly with abortrequested checks?

Reply:
This should be simplified a lot. We have code in other projects that do this more elegant :-)